File: 1/tests/accel/accel.py
#!/usr/bin/evn python

__copyright__ = "Copyright (c) 2011 Slade Maurer, Alexander Sviridenko"

import time

from bb import os
from bb import app
from bb.os.hardware.boards import PropellerDemoBoard

def freefall():
  # freefall is defined as an alternative to static variable in C
  freefall.init_complete = getattr(freefall, 'init_complete', False)
  # Waiting for a messages from h48c device driver
  message = os.get_running_kernel().receive_message('FREEFALL')
  if message and message.get_sender() is h48c.get_name():
    if message.get_command() is h48c.find_command('BBOS_DRIVER_OPEN'):
      print "H48C device has been opened"
      freefall.init_complete = True
      return
    elif message.get_command() is h48c.find_command('H48C_FREEFALL'):
        if message.get_data() == 1:
            print "Free fall detected!"
        elif message.get_data() == 0:
            print "No free fall"
  # Initialization is not complete, send a message to open the driver
  if not freefall.init_complete:
    print "Send open-message to accelerometer device driver"
    os.get_running_kernel().send_message(h48c.get_name(), 
      os.Message("FREEFALL", h48c.find_command('BBOS_DRIVER_OPEN'), None))
  os.get_running_kernel().send_message(h48c.get_name(),
    os.Message("FREEFALL", h48c.find_command('H48C_FREEFALL'), None))
  time.sleep(2)

kernel = os.Kernel()
kernel.set_scheduler(os.StaticScheduler())
kernel.add_thread('FREEFALL', freefall)
h48c = kernel.add_module('bb.os.hardware.drivers.accel.h48c')

accel = app.Process("Accel", kernel)
board = PropellerDemoBoard(processes=[accel])

if app.get_mode() is app.SIMULATION_MODE:
    accel.run()

File: 2/tests/accel/accel.c
/*
 * Copyright (c) 2011 Sladeware LLC
 */

#include <bb/os/config.h>
#include <bb/os/kernel/system.h>
#include <bb/os/hardware/drivers/accel/h48c/core.h>

static struct h48c_pins freefall_pins = {7, 6, 0, 1};
static bool_t init_complete = 0;
static int is_freefall;

void
freefall_runner()
{
  static bbos_message_t message; // message to communicate with accelerometer

  /* Freefall has a new message */
  if (bbos_port_receive(FREEFALL, &message) == BBOS_SUCCESS) {
	/* Try to figure out the command */
	switch (message.id) {
	case BBOS_DRIVER_OPEN:
	  printf("H48C device has been opened\n");
	  init_complete = 1;
	  break;
	  /* Response from the free fall detection */
	case H48C_FREEFALL:
	  if (*(int *)message.data == 1)
	    printf("Free fall detected!\n");
	  break;
	default:
	  bbos_panic("Unknown message");
	  break;
	}
  }

  if (!init_complete) {
	printf("Send open-message to H48C device driver\n");
	message.id = BBOS_DRIVER_OPEN;
	message.data = &freefall_pins;
	bbos_port_send(H48C, &message, FREEFALL);
	return;
  }

  message.id = H48C_FREEFALL;
  message.data = &is_freefall;
  bbos_port_send(H48C, &message, FREEFALL);
}

int
main()
{
  bbos_init();
  bbos_add_thread(FREEFALL, freefall_runner);
  bbos_add_thread(H48C, h48c_runner);
  bbos_start();
  return 0;
}

